### ISSUE TITLE
[Due for payment 2025-02-25] Whisper to admin and auditor when their role is changed to user

### ISSUE THREAD

**Issue by garrettmknight** [Expensify/App#49634]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** N/A
**Reproducible in staging?:** N
**Reproducible in production?:** N
**Expensify/Expensify Issue URL:**
**Issue reported by:** @JmillsExpensify 
**Slack conversation:** https://expensify.slack.com/archives/C06ML6X0W9L/p1727169064223989?thread_ts=1726494273.889699&cid=C06ML6X0W9L

## Problem
When we change a user's role from Admin or Auditor to user we remove them from all workspace chats they had access to, except for their own. Since they've been removed from the #admins room for the workspace, they won't have any context on their role change or why they no longer have access to the chats they once did.

## Solution
Similar to how we handle approvers who are removed from workspace chats ([PR for that here](https://github.com/Expensify/Auth/pull/12328)), we'll send a whisper to admins and auditors in their workspace chat when their role is updated and access to other workspace chats is removed.

**Admin copy for the whisper:**
> [Admin] updated your role in [Workspace Name] from admin to user. You have been removed from all submitter workspace chats except for your own.

**Auditor copy for the whisper:**
> [Admin] updated your role in [Workspace Name] from auditor to user. You have been removed from all submitter workspace chats except for your own.

## Platforms:
All

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Issue Owner</summary>Current Issue Owner: @bfitzexpensify</details>

**Comment by bfitzexpensify** [comment]
Same update - remains a hot pick, waiting to be picked up. Weekly still seems right to me

**Comment by bfitzexpensify** [comment]
Remains a hot pick, waiting to be picked up.

**Comment by blimpich** [comment]
Picking up. Initial investigation shows [this is where we update a modified user's role in Web](https://github.com/Expensify/Web-Expensify/blob/2ef060d654fad078142a7f6cd4d2c1cd4a248289/lib/PolicyAPI.php#L1483). Will continue looking into this tomorrow. Questions I need to answer:
- [ ] do we have the ability to add whisper messages from web? From auth?
- [ ] would it violate 1:1:1 to add code in Web that created these whispers?

**Comment by blimpich** [comment]
To answer my own questions, we have the ability to add whisper messages both from Auth and Web. But, since `UpdateWorkspaceMembersRole` is not 1:1:1 (more so than most commands) its not possible to do this in Auth right now. And while its feasible to do this from Web it would require loading the downgraded User's entire list of chat reports, which makes this flow even more in violation of our 1:1:1 principle.

Maybe there is another engineer who can see something that I can't, but might need to take a raincheck for issue. Going to unassign this from myself for now since I can't prioritize it given how much work it would be to make it comply with 1:1:1.

**Comment by bfitzexpensify** [comment]
Remains a hot pick and available for the taking

**Comment by bfitzexpensify** [comment]
Still waiting on a volunteer, remains low priority

**Comment by bfitzexpensify** [comment]
Same update, waiting on a volunteer

**Comment by tgolen** [comment]
The `UpdateWorkspaceMembersRole` command uses `PolicyAPI::mergeEmployees()` under the hood. I think it would be very simple to make `UpdateWorkspaceMembersRole` call straight to Auth (adding a new `UpdateWorkspaceMembersRole` command to Auth), and having it use the policy merge stuff in Auth that has already been written and in use for quite some time. This would refactor the command to be 1:1:1.

@flodnv @iwiznia do you agree? If so, I can get it moved and implement this whisper I don't think it should take too much work.

**Comment by iwiznia** [comment]
Don't know the full details, but I think as long as we are only changing the role it should be fine. If we are inviting/uninviting then there's probably more to it.

**Comment by flodnv** [comment]
+1, though it would be good to double check. For example, it seems like [this logic would apply if changing the role to admin](https://github.com/Expensify/Auth/blob/b7d574e5ab0ebde294590b01e999d5f95a1631bb/auth/command/SharePolicy.cpp#L161-L164). Perhaps there's more.

**Comment by bfitzexpensify** [comment]
Remains a hot pick - sounds like perhaps we have an idea of a solution @tgolen?

**Comment by tgolen** [comment]
# Weekly Update
- It sounds like @flodnv might have other plans for someone to work on this, so I am not going to start any work on it currently.
- @flodnv feel free to assign this to someone else if you'd like.

**Comment by tgolen** [comment]
# Weekly Update
- No update

### Next steps
- @flodnv take this one when you are ready

**Comment by tgolen** [comment]
# Weekly Update
- Same as last week

**Comment by flodnv** [comment]
@heekinho will work on this!

**Comment by flodnv** [comment]
Some further details about the previous issue can be found here: https://github.com/Expensify/Expensify/issues/422963#issuecomment-2316119567

**Comment by heekinho** [comment]
👍

**Comment by heekinho** [comment]
> To answer my own questions, we have the ability to add whisper messages both from Auth and Web. But, since `UpdateWorkspaceMembersRole` is not 1:1:1 (more so than most commands) its not possible to do this in Auth right now. And while its feasible to do this from Web it would require loading the downgraded User's entire list of chat reports, which makes this flow even more in violation of our 1:1:1 principle.

I believe I have this same take when looking at it. Doing that from Web would be possible but it does violate the principles.
I even implemented a draft PR but I think it's not the ideal solution.


> The `UpdateWorkspaceMembersRole` command uses `PolicyAPI::mergeEmployees()` under the hood. I think it would be very simple to make `UpdateWorkspaceMembersRole` call straight to Auth (adding a new `UpdateWorkspaceMembersRole` command to Auth), and having it use the policy merge stuff in Auth that has already been written and in use for quite some time. This would refactor the command to be 1:1:1.

About this bit, the `PolicyAPI::mergeEmployees()` have quite some code, but I couldn't find a clear mapping on Auth. Can you provide some pointers about that in Auth, @tgolen @flodnv.

**Comment by flodnv** [comment]
Let's chat about this here: https://expensify.slack.com/archives/C086VHWPPUM/p1738095538411439

**Comment by heekinho** [comment]
Opened a PR for it. I have some questions:

1. When changing roles I'm receiving the welcome message again (which I believe is a mistake. I'll double check but since it's not actually related to this issue, I'll open another one if that's the case). Which makes the result look like this, but of course this change is only responsible for the second message:

![Image](https://github.com/user-attachments/assets/fce8a7fa-8c39-4ce4-bef4-b4422d2a83c3)

It looks ok, but if we fix that behavior, and we get the sole whisper, we would get a result like this:

![Image](https://github.com/user-attachments/assets/7612dd89-4071-4943-b10a-19282f72fb18)
Which looks nice but there's a unneded repetition of the name of the admin.

From an UX perspective, what about changing the copy to "Your role in [Workspace name] was changed from admin to user. [...]"

**Comment by flodnv** [comment]
> When changing roles I'm receiving the welcome message again (which I believe is a mistake. I'll double check but since it's not actually related to this issue, I'll open another one if that's the case)

Let's get to the bottom of why this is happening. Does this happen without your changes (i.e., if all your repos are on `main`)?

>From an UX perspective, what about changing the copy to "Your role in [Workspace name] was changed from admin to user. [...]"

Looking at [what we do in the other scenario](https://github.com/Expensify/App/pull/48508#issuecomment-2339015651), it seems lie we just want to remove `John Silva ` from what you have in your screenshot

**Comment by heekinho** [comment]
> Let's get to the bottom of why this is happening. Does this happen without your changes (i.e., if all your repos are on main)?

Yes. It does happen for without the changes. It's on any role change. Doesn't even need to be a demotion.
I see there's a flag to `skipWelcomeMessage` so it's probably not being setup somewhere.

> Looking at [what we do in the other scenario](https://github.com/Expensify/App/pull/48508#issuecomment-2339015651), it seems lie we just want to remove `John Silva ` from what you have in your screenshot

Just to confirm, it would look like this, right?

![Image](https://github.com/user-attachments/assets/232fbbe1-ff83-46e8-aab4-72c751cf6ac2)

I'm seeing the changes for this other scenario and it hints I will have to change it to be more robust, probably following the initial plan with the ReportActions and have some frontend work to deal with that (including i18n), is that correct, @flodnv? I'll wait to see you first look on PR and proceed to change if that's the case.

**Comment by flodnv** [comment]
> Yes. It does happen for without the changes. It's on any role change. Doesn't even need to be a demotion.

Interesting, let me get you added to a channel where you can report bugs.

> Just to confirm, it would look like this, right?

> probably following the initial plan with the ReportActions and have some frontend work to deal with that (including i18n), is that correct, @flodnv?

Yes, I think that's correct 👍

**Comment by heekinho** [comment]
> Interesting, let me get you added to a channel where you can report bugs.
Nice, @flodnv. I will test and collect the screen captures to properly report it on the channel tomorrow.

> Yes, I think that's correct 👍
Ok, I'll take a look at the localization code and improve the PR to provide better support for it 😀

**Comment by heekinho** [comment]
@flodnv Updated the Auth PR, and started drafting the above PRs

**Comment by flodnv** [comment]
@iwiznia can you help to review @heekinho's PRs please?

**Comment by heekinho** [comment]
@flodnv, I'm finishing setting up all environments and tests to complete the checklist for the App PR

**Comment by heekinho** [comment]
@flodnv I finished the environment setup, tested in all environments and adjusted Auth to properly deliver the onyx update.
I'll wrap up by finishing the checklist in App next.

**Comment by heekinho** [comment]
Only the Auth part to go now, @flodnv. 😃 
I guess we only need to wait QA on the App PR to unhold the Auth and complete here, right?

**Comment by iwiznia** [comment]
Yeah, we need to wait for the App PR to get deployed to prod, then we can unhold the auth PR.

**Comment by flodnv** [comment]
Nice!!

**Comment by bfitzexpensify** [comment]
@heekinho please complete the step here https://github.com/Expensify/App/issues/49634#issuecomment-2664714145

**Comment by bfitzexpensify** [comment]
Those look good to me, thanks @heekinho. Opened https://github.com/Expensify/Expensify/issues/474124. Closing this out.

**Comment by flodnv** [comment]
Yep, thanks!
